docs/performance/tune-performance.md

```diff
@@ -331,6 +331,126 @@ cudaProviderOptions.UpdateOptions(providerOptionsDict);
+  alloc_->Free(ptr);
+}
+
+const Ort::Allocator* alloc_;
+};
+  
+// Enable cuda graph in cuda provider option.
+OrtCUDAProviderOptionsV2* cuda_options = nullptr;
+api.CreateCUDAProviderOptions(&cuda_options);
+std::unique_ptr<OrtCUDAProviderOptionsV2, decltype(api.ReleaseCUDAProviderOptions)> rel_cuda_options(cuda_options, api.ReleaseCUDAProviderOptions);
+std::vector<const char*> keys{"enable_cuda_graph"};
+std::vector<const char*> values{"1"};
+api.UpdateCUDAProviderOptions(rel_cuda_options.get(), keys.data(), values.data(), 1);
+
+Ort::SessionOptions session_options;
+api.SessionOptionsAppendExecutionProvider_CUDA_V2(static_cast<OrtSessionOptions*>(session_options), rel_cuda_options.get();
+
+
+// Create IO bound inputs and outputs.
+Ort::Session session(*ort_env, L"matmul_2.onnx", session_options);
+Ort::MemoryInfo info_cuda("Cuda", OrtAllocatorType::OrtArenaAllocator, 0, OrtMemTypeDefault);
+Ort::Allocator cuda_allocator(session, info_cuda);
+
+const std::array<int64_t, 2> x_shape = {3, 2};
+std::array<float, 3 * 2> x_values = {1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f};
+auto input_data = std::unique_ptr<void, CudaMemoryDeleter>(cuda_allocator.Alloc(x_values.size() * sizeof(float)),
+                                                             CudaMemoryDeleter(&cuda_allocator));
+cudaMemcpy(input_data.get(), x_values.data(), sizeof(float) * x_values.size(), cudaMemcpyHostToDevice);
+
+// Create an OrtValue tensor backed by data on CUDA memory
+Ort::Value bound_x = Ort::Value::CreateTensor(info_cuda, reinterpret_cast<float*>(input_data.get()), x_values.size(),
+                                              x_shape.data(), x_shape.size());
+
+const std::array<int64_t, 2> expected_y_shape = {3, 2};
+std::array<float, 3 * 2> expected_y = {1.0f, 4.0f, 9.0f, 16.0f, 25.0f, 36.0f};
+auto output_data = std::unique_ptr<void, CudaMemoryDeleter>(cuda_allocator.Alloc(expected_y.size() * sizeof(float)),
+                                                              CudaMemoryDeleter(&cuda_allocator));
+
+// Create an OrtValue tensor backed by data on CUDA memory
+Ort::Value bound_y = Ort::Value::CreateTensor(info_cuda, reinterpret_cast<float*>(output_data.get()),
+                                                expected_y.size(), expected_y_shape.data(), expected_y_shape.size());
+
+Ort::IoBinding binding(session);
+binding.BindInput("X", bound_x);
+binding.BindOutput("Y", bound_y);
+
+// One regular run for necessary memory allocation and graph capturing
+session.Run(Ort::RunOptions(), binding);
+
+// After capturing, CUDA graph replay happens from this Run onwards
+session.Run(Ort::RunOptions(), binding);
+
+// Update input and then replay CUDA graph with the updated input
+x_values = {10.0f, 20.0f, 30.0f, 40.0f, 50.0f, 60.0f};
+cudaMemcpy(input_data.get(), x_values.data(), sizeof(float) * x_values.size(), cudaMemcpyHostToDevice);
+session.Run(Ort::RunOptions(), binding);
+```
+
+* C#
+
+Will be supported in future releases
```

Comment:
I realized that the C# OrtValue needs something like the Python structure's `update_inplace()` that takes an `OrtExternalAllocation` or something to that effect to support cuda graphs from C#. I will add an enhancement if there is enough request for this feature in C#.